python/hidet/graph/ops/matmul/matmul_f32_x86.py

```diff
@@ -827,36 +812,58 @@ def gemm_5th_loop(
-
-                init_thr(packa_thrcomm_barrier_sense, packa_thrcomm_threads_arrived, loop3_nways)
-                init_thr(packb_thrcomm_barrier_sense, packb_thrcomm_barrier_threads_arrived, loop5_nways)
+                a_ptr = cast(a, ~float32)
+                b_ptr = cast(b, ~float32)
+                c_ptr = cast(c, ~float32)
 
                 parallel_attr = 'p' + str(nthreads)
                 # The outermost loop spawning threads
-                for tidx in grid(nthreads, attrs=parallel_attr):
-                    tid_5th_loop = tidx
-                    work_id_5th_loop = tid_5th_loop // (nthreads // loop5_nways)
-                    comm_id_5th_loop = tid_5th_loop
-
-                    gemm_5th_loop(a, b, c, work_id_5th_loop, comm_id_5th_loop)
-
-            assert isinstance(matmul_kernel_x86_v3, hidet.ir.Function)
-            # matmul_kernel_x86_v3.kind = "cpu_kernel"
+                for batch in range(batch_size):
+                    init_thr(packa_thrcomm_barrier_sense, packa_thrcomm_threads_arrived, loop3_nways)
+                    init_thr(packb_thrcomm_barrier_sense, packb_thrcomm_barrier_threads_arrived, loop5_nways)
+                    # Iterate through the batch dimension, and for each batch,
+                    # locate the corresponding a, b, and c matrices, and then call the single matmul kernel
+                    a_matrix_size = m_size * k_size
+                    b_matrix_size = k_size * n_size
+                    c_matrix_size = m_size * n_size
+                    a_matrix = as_tensor_pointer(a_ptr + (batch * a_matrix_size), dtype=float32, shape=[m_size, k_size])
+                    b_matrix = as_tensor_pointer(b_ptr + (batch * b_matrix_size), dtype=float32, shape=[k_size, n_size])
+                    c_matrix = as_tensor_pointer(c_ptr + (batch * c_matrix_size), dtype=float32, shape=[m_size, n_size])
+                    for tidx in grid(nthreads, attrs=parallel_attr):
+                        tid_5th_loop = tidx
+                        work_id_5th_loop = tid_5th_loop // (nthreads // loop5_nways)
+                        comm_id_5th_loop = tid_5th_loop
+
+                        gemm_5th_loop(a_matrix, b_matrix, c_matrix, work_id_5th_loop, comm_id_5th_loop)
+
+            assert isinstance(matmul_kernel_x86, hidet.ir.Function)
+            # matmul_kernel_x86.kind = "cpu_kernel"
             ir_module = module.ir_module()
             return ir_module
 
 
 class Matmulx86Op(Operator):
     def __init__(self, a: Tensor, b: Tensor):
-        if not (len(a.shape) == len(b.shape) == 2 and a.shape[1] == b.shape[0]):
-            raise ValueError('Matrix multiplication: incompatible sizes: {} and {}'.format(a.shape, b.shape))
+        # if not (len(a.shape) == len(b.shape) == 2 and a.shape[1] == b.shape[0]):
+        #     raise ValueError('Matrix multiplication: incompatible sizes: {} and {}'.format(a.shape, b.shape))
+        if not (
+            len(a.shape) == len(b.shape) == 3
```

Comment:
Can we use the same template to support matmul like:
- [12, 1024, 1024] @ [1024, 1024]
- [12, 1024, 1024] @ [1, 1024, 1024]
- [1024, 1024] @ [4, 5, 1024, 1024]

You can have a look at https://github.com/hidet-org/hidet/blob/main/python/hidet/graph/ops/matmul/batch_matmul.py as a reference.